✨ Set up Copilot instructions (#31)
cecimerelo/BraveNewAlgorithm.jl: .github/copilot-instructions.md

# GitHub Copilot Instructions for BraveNewAlgorithm.jl

## Repository Overview

> BraveNewAlgorithm.jl is a Julia metaheuristic optimization algorithm inspired by Aldous Huxley's "Brave New World". It uses a caste-based population system with five distinct castes (ALPHA, BETA, GAMMA, DELTA, EPSILON) to maintain exploration/exploitation balance.

Reviewer: ```suggestion
BraveNewAlgorithm.jl is a Julia metaheuristic optimization algorithm inspired by Aldous Huxley's "Brave New World". Population is divided in different groups, which we call "castes" imitating the terminology of the novel; different groups will have different ways of  applying the usual genetic operators: selection, crossover and mutation, all of them inspired on the novel. The intention of this is to keep the exploration/exploitation balance, that is, the amount of new solutions generated *exploring* the space vs the *exploitation* or selection of the best candidates following a specific path to maximization.
```

│   ├── BraveNewAlgorithm.jl     # Main module file
│   ├── brave_new_algorithm.jl   # Main algorithm implementation
│   ├── commons.jl               # Common data structures
│   ├── utils.jl                 # Utility functions
│   ├── individual/              # Individual and fitness structures
│   ├── methods/                 # Algorithm methods (hatchery, evolution, etc.)
│   └── operators/               # Genetic operators (crossover, mutation, selection)
├── examples/                     # Usage examples and demos
│   ├── setup.jl                # Dependency setup script
│   ├── quickstart.jl           # Quick start example
│   ├── simple_test.jl          # Basic functionality test
│   └── basic_usage_example.jl  # Comprehensive examples
├── test/                        # Test suite
│   ├── runtests.jl             # Main test runner
│   ├── methods/                # Method tests
│   ├── operators/              # Operator tests
│   └── Config Files/           # Test configuration files
├── data/                        # Data files and results
├── Project.toml                 # Julia package dependencies
├── Manifest.toml               # Locked dependency versions

> ├── Makefile                    # Build automation

Reviewer: ```suggestion
├── Makefile                    # Build and test automation
```

├── README.md                   # Main documentation
└── USAGE.md                    # Detailed usage guide
```

## Development Environment Setup

### Prerequisites
- Julia 1.4+ (tested with Julia 1.10.10)
- Git for version control

### Quick Setup
```bash
# Clone repository
git clone https://github.com/cecimerelo/BraveNewAlgorithm.jl.git
cd BraveNewAlgorithm.jl

# Install dependencies (choose one method)
julia examples/setup.jl              # Recommended: user-friendly setup
make instantiate                     # Alternative: Makefile
julia --project=. -e "using Pkg; Pkg.instantiate()"  # Manual
```

### Build Process
- **Dependency Installation**: ~3-5 minutes (varies by network)
- **Package Precompilation**: ~2-4 minutes
- **Test Suite**: ~30-60 seconds

### Validation Commands
```bash
# Test installation
make tests                           # Run full test suite (~30-60 seconds)

> julia examples/simple_test.jl       # Basic functionality test (~10 seconds)

Reviewer: ```suggestion
julia examples/simple_test.jl       # Basic functionality test (~10 seconds) this does not work for the time being
```

    Dict("ALPHA" => 10, "BETA" => 20, "GAMMA" => 30, "DELTA" => 25, "EPSILON" => 15),
    Dict("ALPHA" => 5, "BETA" => 8, "GAMMA" => 12, "DELTA" => 15, "EPSILON" => 20)
)

# Fitness function (Sphere function)
fitness_func = FitnessFunction(BlackBoxOptimizationBenchmarking.BBOBFunctions[1])

# Model and execution
model = PopulationModel(config, fitness_func, (-5.0, 5.0), 
                       (element, ff) -> element >= ff.fitness_function.f_opt + 1e-8)
generation, results = brave_new_algorithm(model)

println("Best fitness: $(minimum(results.F_Values))")
```

### Configuration File Usage
```julia
# Create config.json
config_dict = Dict(
    "CHROMOSOME_SIZE" => 10,
    "POPULATION_SIZE" => 50,
    "MAX_GENERATIONS" => 200,
    "POPULATION_PERCENTAGE" => Dict("ALPHA" => 15, "BETA" => 25, "GAMMA" => 25, "DELTA" => 20, "EPSILON" => 15),
    "MUTATION_RATE" => Dict("ALPHA" => 3, "BETA" => 6, "GAMMA" => 10, "DELTA" => 15, "EPSILON" => 20)
)

using JSON
open("config.json", "w") do f
    JSON.print(f, config_dict, 4)
end

# Load configuration
config_parameters = read_parameters_file("config.json")
```

## Caste System Guidelines

### Recommended Distributions

> **Exploitation-focused** (for fine-tuning):

Reviewer: ```suggestion
**Exploitation-focused** (for fine-tuning): The "upper" castes get the biggest proportion
```
